### PR TITLE
add deformable conv v1 op and cpu version of deformable conv v2

### PR DESCRIPTION
**Pull request by cjt222**
Deformable convolution v1 is a new method of which convolution, which add offset to every kernel to make reception field bigger and learn to capture irregular object，which is always used in object convolution and segmentation.
Inside, cpu version and gpu version of deformable convolution v1 and cpu version of Deformable convolution v2 are added. Deformable convolution v1 and v2 is used same python API, which decide which version to used by one parameter named modulated.
image

### CODE REVIEW

File: paddle/fluid/operators/deformable_conv_op.cc [hunk 1]
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "paddle/fluid/operators/deformable_conv_op.h"
+#include <memory>
```
Comment (baiyfbupt) [inline paddle/fluid/operators/deformable_conv_op.cc:16]: is this header file needed? if not, consider removing it?

File: paddle/fluid/operators/deformable_conv_cu.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (baiyfbupt) [inline paddle/fluid/operators/deformable_conv_cu.h]: If these kernels are used by CPU and GPU, maybe the naming `deformable_conv_**cu**` for this file is not appropriate?

File: python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+        conv_param = {
+            'stride': self.stride,
+            'pad': self.pad,
+            'dilation': self.dilations
+        }
+
+        input = np.random.random(self.input_size).astype(self.dtype)
+        offset = 10 * np.random.random(self.offset_size).astype(self.dtype)
+        #print(offset)
```
Comment (baiyfbupt) [inline python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py]: Remove these redundant comments

File: python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            'im2col_step': self.im2col_step,
+            'dilations': self.dilations,
+        }
+        self.outputs = {'Output': output}
+        #print("output", output.shape)
+    def has_cuda(self):
+        return core.is_compiled_with_cuda()
+
+    def test_check_output(self):
+        if self.has_cuda():
```
Comment (baiyfbupt) [inline python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py]: Consider removing these  `has_cuda` checks?

Comment (baiyfbupt) [reply python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py]: Same as test_deformable_conv_op.py

File: paddle/fluid/operators/deformable_conv_cu.h (hunk 2 above)
Comment (cjt222) [reply paddle/fluid/operators/deformable_conv_cu.h]: Done

File: python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py (hunk 3 above)
Comment (cjt222) [reply python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py]: Done

File: paddle/fluid/operators/deformable_conv_op.cc (hunk 1 above)
Comment (cjt222) [reply paddle/fluid/operators/deformable_conv_op.cc:16]: this header file is needed for registering cpu file, pre-commit is not passed if remove it

File: python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py (hunk 4 above)
Comment (cjt222) [reply python/paddle/fluid/tests/unittests/test_deformable_conv_v1_op.py]: Done

**Comment by heavengate** [comment]
历史commit有chengjuntao的，这个用户没有assign cla，另外API.spec生成得有问题，请重新生成

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -13198,8 +13198,6 @@ def _get_default_param_initializer():
     pre_bias = helper.create_variable_for_type_inference(dtype)
 
     if modulated:
-        if not isinstance(mask, Variable):
-            raise TypeError("Input Mask of deformable_conv must be Variable")
```
Comment (baiyfbupt) [inline python/paddle/fluid/layers/nn.py]: mask只能是Variable类型的，这里的类型检查是否是必要的？
这个PR可以先merge，后续再提PR加上

**Review by baiyfbupt — DISMISSED** [review]
LGTM

**Review by heavengate — DISMISSED** [review]
LGTM

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (cjt222) [reply python/paddle/fluid/layers/nn.py]: 之后会提PR加上

**Review by nepeplwu — DISMISSED** [review]

**Review by raindrops2sea — APPROVED** [review]
LGTM

**Review by nepeplwu — APPROVED** [review]
LGTM

**Review by heavengate — APPROVED** [review]

**Pull request merged**